parquet-column/src/main/java/org/apache/parquet/internal/column/columnindex/ColumnIndexBuilder.java

```diff
@@ -150,13 +191,164 @@ boolean isNullPage(int pageIndex) {
       return nullPages[pageIndex];
     }
 
-    abstract ByteBuffer getMinValueAsBytes(int pageIndex);
+    /*
```

Comment:
For the public access of `ColumnIndex` all methods use *page index*es. Only package private access methods may use the *array index*. The latter one is required for the filtering implementations to perform better. Because of these performance sensitive calls I cannot see any proper way to have type-safe distinction between the two.
Method names may indicate which index shall be used but long names may also lead to understand the algorithms harder. I've tried to comment the internal logic and the usage of the proper indexing. I one have to modify the related code parts and get confused on the indexing the related unit tests shall catch the problems and the comments shall help to understand the situation. I don't have any better idea to improve this part.